File: app/src/main/java/be/digitalia/fosdem/viewmodels/BookmarksViewModel.java
package be.digitalia.fosdem.viewmodels;

import android.app.Application;
import android.os.AsyncTask;
import android.text.format.DateUtils;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import be.digitalia.fosdem.db.AppDatabase;
import be.digitalia.fosdem.livedata.LiveDataFactory;
import be.digitalia.fosdem.model.Event;

import java.util.List;
import java.util.concurrent.TimeUnit;

public class BookmarksViewModel extends AndroidViewModel {

	// In upcomingOnly mode, events that just started are still shown for 5 minutes
	static final long TIME_OFFSET = 5L * DateUtils.MINUTE_IN_MILLIS;

	private final AppDatabase appDatabase = AppDatabase.getInstance(getApplication());
	private final MutableLiveData<Boolean> upcomingOnly = new MutableLiveData<>();
	private final LiveData<List<Event>> bookmarks = Transformations.switchMap(upcomingOnly,
			upcomingOnly -> {
				if (upcomingOnly == Boolean.TRUE) {
					// Refresh upcoming bookmarks every 2 minutes
					final LiveData<Long> heartbeat = LiveDataFactory.interval(2L, TimeUnit.MINUTES);
					return Transformations.switchMap(heartbeat,
							version -> appDatabase.getBookmarksDao().getBookmarks(System.currentTimeMillis() - TIME_OFFSET));
				}

				return appDatabase.getBookmarksDao().getBookmarks(-1L);
			});

	public BookmarksViewModel(@NonNull Application application) {
		super(application);
	}

	public void setUpcomingOnly(boolean upcomingOnly) {
		final Boolean boxedUpcomingOnly = upcomingOnly;
		if (!boxedUpcomingOnly.equals(this.upcomingOnly.getValue())) {
			this.upcomingOnly.setValue(boxedUpcomingOnly);
		}
	}

	public boolean getUpcomingOnly() {
		return Boolean.TRUE.equals(this.upcomingOnly.getValue());
	}

	public LiveData<List<Event>> getBookmarks() {
		return bookmarks;
	}

	public void removeBookmarks(final long[] eventIds) {
		AsyncTask.SERIAL_EXECUTOR.execute(() -> appDatabase.getBookmarksDao().removeBookmarks(eventIds));
	}
}

File: app/src/main/java/be/digitalia/fosdem/viewmodels/BookmarksViewModel.kt
package be.digitalia.fosdem.viewmodels

import android.app.Application
import android.net.Uri
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import be.digitalia.fosdem.BuildConfig
import be.digitalia.fosdem.alarms.AppAlarmManager
import be.digitalia.fosdem.db.BookmarksDao
import be.digitalia.fosdem.db.ScheduleDao
import be.digitalia.fosdem.flow.SharedFlowContext
import be.digitalia.fosdem.flow.stateFlow
import be.digitalia.fosdem.flow.synchronizedTickerFlow
import be.digitalia.fosdem.flow.versionedResourceFlow
import be.digitalia.fosdem.model.Event
import be.digitalia.fosdem.parsers.ExportedBookmarksParser
import be.digitalia.fosdem.utils.BackgroundWorkScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okio.buffer
import okio.source
import java.time.Instant
import javax.inject.Inject
import kotlin.time.Duration.Companion.minutes

@HiltViewModel
class BookmarksViewModel @Inject constructor(
    private val bookmarksDao: BookmarksDao,
    private val scheduleDao: ScheduleDao,
    private val alarmManager: AppAlarmManager,
    private val application: Application
) : ViewModel() {

    private val hidePastEventsStateFlow = MutableStateFlow<Boolean?>(null)

    @OptIn(ExperimentalCoroutinesApi::class)
    val bookmarks: StateFlow<List<Event>?> = stateFlow(viewModelScope, null) {
        hidePastEventsStateFlow.filterNotNull().flatMapLatest { hidePastEvents ->
            if (hidePastEvents) {
                // Refresh upcoming bookmarks every 2 minutes
                synchronizedTickerFlow(REFRESH_PERIOD)
                    .flatMapLatest {
                        getObservableBookmarks(Instant.now())
                    }
            } else {
                getObservableBookmarks(Instant.EPOCH)
            }
        }
    }

    private fun SharedFlowContext.getObservableBookmarks(minEndTime: Instant): Flow<List<Event>> =
        versionedResourceFlow(bookmarksDao.version) {
            bookmarksDao.getBookmarks(minEndTime)
        }

    var hidePastEvents: Boolean
        get() = hidePastEventsStateFlow.value == true
        set(value) {
            hidePastEventsStateFlow.value = value
        }

    fun removeBookmarks(eventIds: LongArray) {
        BackgroundWorkScope.launch {
            if (bookmarksDao.removeBookmarks(eventIds) > 0) {
                alarmManager.onBookmarksRemoved(eventIds)
            }
        }
    }

    suspend fun readBookmarkIds(uri: Uri): LongArray = withContext(Dispatchers.IO) {
        val parser = ExportedBookmarksParser(
            applicationId = BuildConfig.APPLICATION_ID,
            conferenceId = scheduleDao.getYear().toString()
        )
        checkNotNull(application.contentResolver.openInputStream(uri)).source().buffer().use {
            parser.parse(it)
        }
    }

    companion object {
        private val REFRESH_PERIOD = 2.minutes
    }
}